validator/client/validator.go

```diff
@@ -931,6 +935,79 @@ func (v *validator) logDuties(slot types.Slot, duties []*ethpb.DutiesResponse_Du
 	}
 }
 
+// PrepareBeaconProposer calls the prepareBeaconProposer RPC to set the proposer information such as fee recipient.
+func (v *validator) PrepareBeaconProposer(ctx context.Context, km keymanager.IKeymanager) error {
+	if km == nil {
+		return errors.New("keymanager is nil when calling PrepareBeaconProposer")
+	}
+	feeRecipients, err := v.feeRecipients(ctx, km)
+	if err != nil {
+		return err
+	}
+	if len(feeRecipients) == 0 {
+		log.Warnf("no valid validator indices were found, prepare beacon proposer request fee recipients array is empty")
+		return nil
+	}
+	if _, err := v.validatorClient.PrepareBeaconProposer(ctx, &ethpb.PrepareBeaconProposerRequest{
+		Recipients: feeRecipients,
+	}); err != nil {
+		return err
+	}
+	log.Infoln("Successfully prepared beacon proposer with fee recipient to validator index mapping.")
+	return nil
+}
+
+func (v *validator) feeRecipients(ctx context.Context, km keymanager.IKeymanager) ([]*ethpb.PrepareBeaconProposerRequest_FeeRecipientContainer, error) {
+	var validatorToFeeRecipientArray []*ethpb.PrepareBeaconProposerRequest_FeeRecipientContainer
+	if v.prepareBeaconProposalConfig == nil {
+		return nil, errors.New("no config was provided to set validator fee recipients")
+	}
+	// need to check for new keys
+	pubkeys, err := km.FetchValidatingPublicKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, key := range pubkeys {
+		hexKey := hexutil.Encode(key[:])
+		feeRecipient := fieldparams.Eth1BurnAddressHex
+		validatorIndex := v.pubkeyHexToValidatorIndex[hexKey]
+		// ignore updating fee recipient if validator index is not found
+		if validatorIndex == 0 {
+			resp, err := v.validatorClient.ValidatorIndex(ctx, &ethpb.ValidatorIndexRequest{PublicKey: key[:]})
+			if err != nil {
+				// do a strings contains? to see if the error is a not found error
+				if strings.Contains(err.Error(), "Could not find validator index") {
+					log.Infoln("Could not find validator index for public key %#x not found. "+
+						"Perhaps the validator is not yet active.", hexKey)
+					continue
+				}
+				return nil, err
+			} else {
+				validatorIndex = resp.Index
+				// update the cache for the next time
+				v.pubkeyHexToValidatorIndex[hexKey] = validatorIndex
+			}
+		}
```

Comment:
This part of the code kind of violates the good old single responsibility per function principle. As we are returning feeRecipients, we are also updating the cache and all the good stuff.

An ideal refactor would be to refactor this out of feeRecipient and make feeRecipient accept `validatorIndex` instead of `keymanager.IKeymanager`

Reply:
in order to not have an additional for loop i just extracted the validator index rpc call to its own function, removed the km dependency replacing it with public keys instead